### PR TITLE
luks_device.py: Allow manipulate LUKS containers with label or UUID

### LINKED ISSUE #58973: Support creating LUKS2 container in luks_device module
##### SUMMARY

Ansible by default only creates LUKS1 format devices. Considering LUKS2 extra features and age, I think it would be beneficial to allow users to create LUKS2 devices.

##### ISSUE TYPE
- Feature Idea

##### COMPONENT NAME
luks_device

##### ADDITIONAL INFORMATION

For reliability, one should use either disk UUID's or labels. UUID's are (in my opinion) a hassle when it comes to using them in Ansible.

Enter labels. Labels can be the same across multiple hosts, and will be tied to the right disk/partition, even if sda1 suddenly becomes sdb1 after some hardware reconfiguration.

Now, the LUKS1 format, which Ansible creates by default with the `luks_device` module, does not support labels... Support was added with LUKS2.

It should be noted that, from what I can see, the module *does already support removing keys from LUKS2 format devices*.

```yaml
- name: create LUKS2 container
  luks_device:
    device: "/dev/loop0"
    state: "present"
    keyfile: "/vault/keyfile"
    format: luks2
```

### PR DESCRIPTION
**Pull request by mulatinho**
Per default, on newer versions, cryptsetup creates LUKS containers with VERSION 2 format, but with this change it allow user create containers with `LABEL` identifier as requested on #58973 expliciting `LUKS2` format. It also allow manipulation of LUKS containers with `UUID`

- Allow create a LUKS2 container format with label support
- Allow manipulate (open, close, modify) an LUKS container based on
  both label (LUKS2 format) or UUID (V1 ans V2) instead of using devices only.

This was tested on Fedora Core 30 server and a Slackware 14.2 server.

Fixes: #58973

How to test these new features
```
---
- name: test issue 56872
  gather_facts: no
  hosts: freeipa
  remote_user: root
  tasks:
    - name: create images
      command: dd if=/dev/urandom of=/dev/fedora/test200M bs=100M count=2

    - name: create LUKS container
      luks_device:
        device: "/dev/fedora/test200M"
        state: "present"
        keyfile: "/vault/keyfile"
        label: "personalLabelName"

    - name: open LUKS container
      luks_device:
        state: "opened"
        name: "cryptDiskVol"
        keyfile: "/vault/keyfile"
        label: "personalLabelName"

    - name: close LUKS container
      luks_device:
        state: "closed"
        uuid: ae97ed88-b6eb-40e1-a32e-c2f08d8f0de2
```

Files changed:
- lib/ansible/modules/crypto/luks_device.py

### CODE REVIEW

File: lib/ansible/modules/crypto/luks_device.py
Comment (felixfontein) [inline lib/ansible/modules/crypto/luks_device.py]:
```suggestion
            - "This option allow the user to create a LUKS2 format container with
```

File: lib/ansible/modules/crypto/luks_device.py
Comment (felixfontein) [inline lib/ansible/modules/crypto/luks_device.py]:
```suggestion
              label support, respectively to identifiy the container by label on later usages.
```

File: lib/ansible/modules/crypto/luks_device.py
Comment (felixfontein) [inline lib/ansible/modules/crypto/luks_device.py]: This will collide with #61522. Also, you should use a similar approach as in there, i.e. don't do an `if` with two copies of `self._run_command()`, but collect the options in a list and then finally call `self._run_command()`.

File: lib/ansible/modules/crypto/luks_device.py
Comment (felixfontein) [inline lib/ansible/modules/crypto/luks_device.py]: Is there a reason why you want to remove this? Does this improve compatibility in any way, or is required for LUKS2 support?

File: lib/ansible/modules/crypto/luks_device.py
Comment (felixfontein) [inline lib/ansible/modules/crypto/luks_device.py]:
```suggestion
            error_msg = "Can't find the device based on label, UUID or name. " \
```

File: lib/ansible/modules/crypto/luks_device.py
Comment (felixfontein) [inline lib/ansible/modules/crypto/luks_device.py]:
```suggestion
                        "Please check whether the provided parameters are correct."
```

File: lib/ansible/modules/crypto/luks_device.py
Comment (felixfontein) [inline lib/ansible/modules/crypto/luks_device.py]:
```suggestion
```

File: lib/ansible/modules/crypto/luks_device.py
Comment (felixfontein) [inline lib/ansible/modules/crypto/luks_device.py]:
```suggestion
        version_added: "2.10"
```
(same below)

**Comment by felixfontein** [comment]
You'll also need a [changelog fragment](https://docs.ansible.com/ansible/latest/community/development_process.html#changelogs-how-to) (`minor_changes`).

File: lib/ansible/modules/crypto/luks_device.py
Comment (mulatinho) [reply lib/ansible/modules/crypto/luks_device.py]: Fixed, thanks! ;-)

File: lib/ansible/modules/crypto/luks_device.py
Comment (mulatinho) [reply lib/ansible/modules/crypto/luks_device.py]: Fixed, thanks! ;-)

File: lib/ansible/modules/crypto/luks_device.py
Comment (mulatinho) [reply lib/ansible/modules/crypto/luks_device.py]: Fixed, thanks! ;-)

File: lib/ansible/modules/crypto/luks_device.py
Comment (mulatinho) [reply lib/ansible/modules/crypto/luks_device.py]: Its required for LUKS2 support, if pass `--type luks` it will make LUKS1 container on some versions.

File: lib/ansible/modules/crypto/luks_device.py
Comment (mulatinho) [reply lib/ansible/modules/crypto/luks_device.py]: Fixed, thanks! :)

File: lib/ansible/modules/crypto/luks_device.py
Comment (mulatinho) [reply lib/ansible/modules/crypto/luks_device.py]: Done!

File: lib/ansible/modules/crypto/luks_device.py
Comment (mulatinho) [reply lib/ansible/modules/crypto/luks_device.py]: Done!

File: lib/ansible/modules/crypto/luks_device.py
Comment (mulatinho) [reply lib/ansible/modules/crypto/luks_device.py]: Done! Thanks ;D

**Comment by mulatinho** [comment]
ready_for_review

File: changelogs/fragments/58973_luks_device-add-label-and-uuid-support.yml
Comment (felixfontein) [inline changelogs/fragments/58973_luks_device-add-label-and-uuid-support.yml]:
```suggestion
- luks_device - added support to use UUIDs, and labels with LUKS2 containers
```
UUIDs also work with LUKS1 containers

File: lib/ansible/modules/crypto/luks_device.py
Comment (felixfontein) [inline lib/ansible/modules/crypto/luks_device.py]:
```suggestion
            - "With this option user can identify the LUKS container by UUID.
```

File: lib/ansible/modules/crypto/luks_device.py
Comment (felixfontein) [inline lib/ansible/modules/crypto/luks_device.py:107]:
```suggestion
              label on later usages."
            - "Will only be used on container creation, or when I(device) is not specified."
```

File: lib/ansible/modules/crypto/luks_device.py
Comment (felixfontein) [inline lib/ansible/modules/crypto/luks_device.py]:
```suggestion
            - "Will only be used when I(device) and I(label) are not specified."
```

File: lib/ansible/modules/crypto/luks_device.py
Comment (felixfontein) [inline lib/ansible/modules/crypto/luks_device.py]:
```suggestion
- name: open the LUKS container based on label without device; name it "mycrypt"
```
You don't specify `uuid` here.

File: lib/ansible/modules/crypto/luks_device.py
Comment (felixfontein) [inline lib/ansible/modules/crypto/luks_device.py]:
```suggestion
- name: close container based on UUID
```

File: lib/ansible/modules/crypto/luks_device.py
Comment (felixfontein) [inline lib/ansible/modules/crypto/luks_device.py]:
```suggestion
        query = '%s=%s' % ('UUID', uuid)
```
Since single quotes are used everywhere else.

File: lib/ansible/modules/crypto/luks_device.py
Comment (felixfontein) [inline lib/ansible/modules/crypto/luks_device.py]: Same here.

File: lib/ansible/modules/crypto/luks_device.py
Comment (felixfontein) [inline lib/ansible/modules/crypto/luks_device.py]: Instead of replacing `command`, why not extend it? This way it is very hard to extend this code.

File: lib/ansible/modules/crypto/luks_device.py
Comment (felixfontein) [reply lib/ansible/modules/crypto/luks_device.py]: According to the man page, `luks` (the default value) opens both LUKS1 and LUKS2 containers; see here: http://man7.org/linux/man-pages/man8/cryptsetup.8.html#BASIC_COMMANDS

Do you know of a system where `--type luks` will not open LUKS2 containers? This is about opening, not creating.

File: lib/ansible/modules/crypto/luks_device.py
Comment (felixfontein) [reply lib/ansible/modules/crypto/luks_device.py]: Anyway, it's probably best to wait until #61603 is merged and then change this.

**Comment by felixfontein** [comment]
I've merged #61522, feel free to rebase any when you want :)

File: lib/ansible/modules/crypto/luks_device.py
Comment (mulatinho) [reply lib/ansible/modules/crypto/luks_device.py]: You are right sir, my bad!

File: lib/ansible/modules/crypto/luks_device.py
Comment (mulatinho) [reply lib/ansible/modules/crypto/luks_device.py]: I took your suggestion, thanks!

File: lib/ansible/modules/crypto/luks_device.py
Comment (mulatinho) [reply lib/ansible/modules/crypto/luks_device.py]: Done, thanks!

File: lib/ansible/modules/crypto/luks_device.py
Comment (mulatinho) [reply lib/ansible/modules/crypto/luks_device.py]: Done, thanks!

File: lib/ansible/modules/crypto/luks_device.py
Comment (mulatinho) [reply lib/ansible/modules/crypto/luks_device.py]: Done, thanks!

File: lib/ansible/modules/crypto/luks_device.py
Comment (mulatinho) [reply lib/ansible/modules/crypto/luks_device.py]: You are correct, thanks!

File: lib/ansible/modules/crypto/luks_device.py
Comment (mulatinho) [reply lib/ansible/modules/crypto/luks_device.py]: Done, thanks!

File: lib/ansible/modules/crypto/luks_device.py
Comment (mulatinho) [reply lib/ansible/modules/crypto/luks_device.py]: Done, thanks!

File: lib/ansible/modules/crypto/luks_device.py
Comment (mulatinho) [reply lib/ansible/modules/crypto/luks_device.py:107]: Done, thanks!

File: changelogs/fragments/58973_luks_device-add-label-and-uuid-support.yml
Comment (mulatinho) [reply changelogs/fragments/58973_luks_device-add-label-and-uuid-support.yml]: Done, thanks!

**Comment by mulatinho** [comment]
> I've merged #61522, feel free to rebase any when you want :)

Cool! Thanks for the info, I will do soonly. :D

**Comment by mulatinho** [comment]
ready_for_review

File: lib/ansible/modules/crypto/luks_device.py
Comment (felixfontein) [reply lib/ansible/modules/crypto/luks_device.py]: No problem. In fact, removing it shouldn't hurt, since `luks` is the default. But I guess it's better to keep it, just in case somewhen in the future, the default changes.

File: lib/ansible/modules/crypto/luks_device.py
Comment (felixfontein) [inline lib/ansible/modules/crypto/luks_device.py]:
```suggestion
              with label support, respectively to identify the container by
```

File: lib/ansible/modules/crypto/luks_device.py
Comment (felixfontein) [inline lib/ansible/modules/crypto/luks_device.py]:
```suggestion
```

File: lib/ansible/modules/crypto/luks_device.py
Comment (felixfontein) [inline lib/ansible/modules/crypto/luks_device.py]: Fun fact: when I run this as my user, it returns nothing. When I run it as root, it works. On the other hand, `blkid --uuid <UUID>` also works when I run it as my user.

So why not use
```suggestion
        result = self._run_command([self._blkid_bin, '--uuid', query])
```
instead? That seems to be simpler. Also, you don't need any postprocessing, since it returns only the device name and nothing else.

File: lib/ansible/modules/crypto/luks_device.py
Comment (felixfontein) [inline lib/ansible/modules/crypto/luks_device.py]: Probably the same with `--label` instead of `--uuid` here.

File: lib/ansible/modules/crypto/luks_device.py
Comment (felixfontein) [inline lib/ansible/modules/crypto/luks_device.py]: `blkid` is not always installed (for example Debian 9 and 10 don't necessarily include it). How about not requiring it (i.e. use `False`), and checking whether `self._blkid_bin` contains a result when you actually need it, and if then it doesn't, fail the module? Then `blkid` is not a requirement to run the module at all (right now it is).

Comment (felixfontein) [reply lib/ansible/modules/crypto/luks_device.py]: You can also move this line into `get_device_by_uuid()` and `get_device_by_label()`; then Ansible will fail for you if it cannot find the command, and it will only search for it when we actually need it.

File: lib/ansible/modules/crypto/luks_device.py
Comment (mulatinho) [reply lib/ansible/modules/crypto/luks_device.py]: Nice, much better sir, thanks ;-D

File: lib/ansible/modules/crypto/luks_device.py
Comment (mulatinho) [reply lib/ansible/modules/crypto/luks_device.py]: Nice, much better sir, thanks ;-D

File: lib/ansible/modules/crypto/luks_device.py
Comment (mulatinho) [reply lib/ansible/modules/crypto/luks_device.py]: Hmmm, I choose the second option, thanks for the tip!

**Comment by mulatinho** [comment]
ready_for_review

File: lib/ansible/modules/crypto/luks_device.py
Comment (felixfontein) [inline lib/ansible/modules/crypto/luks_device.py]:
```suggestion
    - "blkid (when I(label) or I(uuid) options are used)"
```

File: lib/ansible/modules/crypto/luks_device.py
Comment (felixfontein) [inline lib/ansible/modules/crypto/luks_device.py]: While we're at it:
```suggestion
    - "wipefs (when I(state) is C(absent))"
```

File: test/units/modules/crypto/test_luks_device.py
Comment (felixfontein) [inline test/units/modules/crypto/test_luks_device.py]: Why did this value change?

File: test/units/modules/crypto/test_luks_device.py
Comment (felixfontein) [inline test/units/modules/crypto/test_luks_device.py]:
```suggestion
        assert expected == "exception"
```
If `expected` is `False`, an exception should not be accepted.

File: test/units/modules/crypto/test_luks_device.py
Comment (felixfontein) [inline test/units/modules/crypto/test_luks_device.py]:
```suggestion
        assert expected == "exception"
```

File: test/units/modules/crypto/test_luks_device.py
Comment (felixfontein) [inline test/units/modules/crypto/test_luks_device.py]:
```suggestion
        assert expected == "exception"
```

File: test/units/modules/crypto/test_luks_device.py
Comment (felixfontein) [inline test/units/modules/crypto/test_luks_device.py]:
```suggestion
        assert expected == "exception"
```

File: test/units/modules/crypto/test_luks_device.py
Comment (felixfontein) [inline test/units/modules/crypto/test_luks_device.py]:
```suggestion
        assert expected == "exception"
```

**Review by felixfontein — COMMENTED** [review]
The module code looks good (except two minor details, only one of them related to this PR). I have some questions about the unit tests, though.

**Comment by mulatinho** [comment]
ready_for_review

File: lib/ansible/modules/crypto/luks_device.py
Comment (felixfontein) [inline lib/ansible/modules/crypto/luks_device.py]: Almost ;)
```suggestion
    - "wipefs (when I(state) is C(absent))"
```

Comment (mulatinho) [reply lib/ansible/modules/crypto/luks_device.py]: LOL, laughing to not cry :P

**Comment by mulatinho** [comment]
ready_for_review

File: lib/ansible/modules/crypto/luks_device.py
Comment (felixfontein) [reply lib/ansible/modules/crypto/luks_device.py]: happens ;)

**Comment by felixfontein** [comment]
It looks like you pushed the wrong version, the current version undoes a lot of things from the previous pushes.

File: test/units/modules/crypto/test_luks_device.py
Comment (mulatinho) [reply test/units/modules/crypto/test_luks_device.py]: I was making some tests and forget to clean this line, these tests are based on a lot of "fake mock results"

> It looks like you pushed the wrong version, the current version undoes a lot of things from the previous pushes.

oh man ;/ sorry, let me fix that

**Comment by mulatinho** [comment]
> It looks like you pushed the wrong version, the current version undoes a lot of things from the previous pushes.

My bad, two computers, a lot of rebases, I revised last versions and seems its OK now.

ready_for_review

File: lib/ansible/modules/crypto/luks_device.py
Comment (felixfontein) [inline lib/ansible/modules/crypto/luks_device.py]:
```suggestion
        result = self._run_command([self._blkid_bin, '--uuid', uuid])
```
According to `man blkid` on my system, it is `--uuid` and not `--uid`.

File: lib/ansible/modules/crypto/luks_device.py
Comment (felixfontein) [inline lib/ansible/modules/crypto/luks_device.py]: The
```suggestion
        uuid = self._module.params['uuid']
```
you had here before is also fine. `AnsibleModule` makes sure that all options will be there; if the user didn't specify them, you'll get `None`.

**Comment by felixfontein** [comment]
> > It looks like you pushed the wrong version, the current version undoes a lot of things from the previous pushes.
> 
> My bad, two computers, a lot of rebases, I revised last versions and seems its OK now.

That's why it's best not to rebase, but to add new commits :)

File: lib/ansible/modules/crypto/luks_device.py
Comment (mulatinho) [reply lib/ansible/modules/crypto/luks_device.py]: I got some errors without the .get method on `test_luks_device.py`, not on this place exactly. dont know why but they are there, well, I change this line and the another on label.

File: lib/ansible/modules/crypto/luks_device.py
Comment (mulatinho) [reply lib/ansible/modules/crypto/luks_device.py]: > > > It looks like you pushed the wrong version, the current version undoes a lot of things from the previous pushes.
> > 
> > 
> > My bad, two computers, a lot of rebases, I revised last versions and seems its OK now.
> 
> That's why it's best not to rebase, but to add new commits :)

I always tought that create a lot of commits was not a good idea, but some of the rebases made here was not cool also, lesson is: dont write code with a late pregnant wife and with your kid of 2 wanting to play, wait for a calm day and write these on the right way (thanks God for all that). But its always the coder not the family behind him :D

sorry for made you waste your time, hope now, everything is OK.

**Comment by mulatinho** [comment]
> > > It looks like you pushed the wrong version, the current version undoes a lot of things from the previous pushes.
> > 
> > 
> > My bad, two computers, a lot of rebases, I revised last versions and seems its OK now.
> 
> That's why it's best not to rebase, but to add new commits :)

> 
> 
>         It looks like you pushed the wrong version, the current version undoes a lot of things from the previous pushes.
> 
>     My bad, two computers, a lot of rebases, I revised last versions and seems its OK now.
> 
> That's why it's best not to rebase, but to add new commits :)

I always tought that create a lot of commits was not a good idea, but some of the rebases made here was not cool also, lesson is: dont write code with a late pregnant wife and with your kid of 2 wanting to play, wait for a calm day and write these on the right way (thanks God for all that). But its always the coder not the family behind him :D

sorry for made you waste your time, hope now, everything is OK.

**Comment by mulatinho** [comment]
ready_for_review

**Review by felixfontein — APPROVED** [review]
LGTM

**Comment by felixfontein** [comment]
@mulatinho thank you very much for implementing this! :)

**Pull request merged**
